pageserver/src/walreceiver.rs

```diff
@@ -27,56 +27,55 @@ use crate::config::PageServerConf;
 use crate::task_mgr::WALRECEIVER_RUNTIME;
 
 use anyhow::{ensure, Context};
-use etcd_broker::Client;
-use itertools::Itertools;
+use neon_broker::neon_broker_proto::neon_broker_client::NeonBrokerClient;
 use once_cell::sync::OnceCell;
 use std::future::Future;
 use tokio::sync::watch;
+use tonic::transport::Channel;
 use tracing::*;
-use url::Url;
 
 pub use connection_manager::spawn_connection_manager_task;
 
-static ETCD_CLIENT: OnceCell<Client> = OnceCell::new();
+static BROKER_CLIENT: OnceCell<NeonBrokerClient<Channel>> = OnceCell::new();
 
 ///
-/// Initialize the etcd client. This must be called once at page server startup.
+/// Initialize the broker client. This must be called once at page server startup.
 ///
-pub async fn init_etcd_client(conf: &'static PageServerConf) -> anyhow::Result<()> {
-    let etcd_endpoints = conf.broker_endpoints.clone();
+pub async fn init_broker_client(conf: &'static PageServerConf) -> anyhow::Result<()> {
+    let broker_endpoints = conf.broker_endpoints.clone();
     ensure!(
-        !etcd_endpoints.is_empty(),
-        "Cannot start wal receiver: etcd endpoints are empty"
+        !broker_endpoints.is_empty(),
+        "Cannot start wal receiver: broker endpoints are empty"
```

Comment:
Trying to understand what's up with the "empty string pattern" and why cannot we avoid it?
So far, my understanding is we're not going to work without broker, since no wall arrives and we do not have the readonly replicas still.
When we will, they might need some broker sync either, as we've discussed in a recent storage sync RFC.

So, feels like we don't really need to turn the broker off and there should be a hard error if it's missing? If so, then an empty string is not really nice IMO and we should always have some url and try to reconnect it endlessly.

If it's not true and we might want to start storage binaries without the broker, we can alter the toml parsing code and make the `broker_url(s)` non-mandatory? So we parse it into some `Option<Url>` or `None` and react accordingly.

If that's not working for some reason and we have to parse/pass some value related to the endpoint into config, I'd vote for keeping an array then, since we could pass an empty array instead of an empty string.

Reply:
> my understanding is we're not going to work without broker, since no wall arrives and we do not have the readonly replicas still.

Generally true, but you might imagine a pageserver serving read only replica at fixed LSN, then WAL stream is not needed.

> non-mandatory? So we parse it into some Option<Url> or None and react accordingly.

Yes, but in this case default value localhost:50051 seems to me more useful than None -- but there is no way to explicitly pass None except as empty string (in which case you don't need Option).

Reply:
> but there is no way to explicitly pass None except as empty string

I was imagining, that if we don't pass anything, that would be the real `None`, but I think I am spending your time on this minor thing for too much and not understanding something.

So, if my proposals don't work, let's move on 🙂